library/src/main/java/SpecTreeComponentSelector.java

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Stack;
+
+public class SpecTreeComponentSelector {
+
+  /**
+   * Given a spec tree, returns only the relevant component refs for the paths present.
+   *
+   * @param spec a spec tree, which has typically been already filtered and might have unnecessary
+   *     components
+   * @return the components which are relevant to paths in {@code spec}
+   */
+  public static LinkedHashMap<String, Object> getRelevantComponents(
+      LinkedHashMap<String, Object> spec) throws UnionConflictException, UnexpectedTypeException {
+
+    var keypaths = new HashSet<String>();
+    addRefsInSubtreeToKeypaths(
+        ObjectCaster.castObjectToStringObjectMap(spec.get("paths")), keypaths);
+
+    var outputComponents = new LinkedHashMap<String, Object>();
+
+    int size = 0;
+    // as soon as keypaths.size() == size, terminate because no new refs need to be looked for.
+    while (keypaths.size() != size) {
```

Comment:
Little bit of code smell here. You have a loop where the exit conditional is depending on something external to change the condition and the variable is being passed to that external method by reference.